### ISSUE TITLE
hyperref citecolor option yields inconsistent colors on brackets and commas.

### ISSUE THREAD

**Issue by simon-carlstedt** [josephwright/beamer#671]
\documentclass[
  xcolor={svgnames},
  hyperref={colorlinks=true,linkcolor=orange,citecolor=blue,urlcolor=blue,bookmarksopen=true}
  ]{beamer}
\mode<presentation>

E.g., citecolor=blue yields author, year and open bracket as blue , but closed bracket as black, i.e.:

\citet{smith1776} => Smith (1776) with:
Blue: 'Smith', '(' and '1776'
Black: ')'

With \citep{smith1776} => (Smith, 1776) the 'Smith', '1776' are blue, '(' and ')' are black, but ',' (the comma in the middle) is blue. 

In other document types author and year are blue; open and closed brackets and commas are black.

**Comment by samcarter** [comment]
@simon-carlstedt  Can you make a minimal working example so we see which bibliography tool you use and which style?

**Comment by simon-carlstedt** [comment]
%\documentclass{beamer}

\documentclass[
xcolor={svgnames},
hyperref={colorlinks=true,linkcolor=orange,citecolor=blue,urlcolor=blue,bookmarksopen=true}
]{beamer}
\mode<presentation>

\usepackage{amsmath}%
\usepackage{amsfonts}%
\usepackage{amssymb}%
\usepackage{graphicx}
\usepackage{amsthm}
\usepackage{mathrsfs}
\usepackage{array}
\usepackage{bm}
\usepackage{amsbsy}
\usepackage{tikz}
\usetikzlibrary{positioning,shapes,arrows}
\usepackage[square,comma,super,numbers,sort&compress]{natbib}
\usepackage{breakcites}
\usepackage{natbib}
\bibpunct{(}{)}{;}{a}{,}{,} % preamble for output, first two are bracket type, i.e. round
\setlength{\bibsep}{0pt} % Make reference list compact
\usepackage{tikz}
\usetikzlibrary{shapes,backgrounds,positioning,arrows}
\usetikzlibrary{chains,fit}
\usepackage{tikz}
\usepackage{verbatim}
\usepackage{rotating}
\usepackage{float}
\usepackage{pgfplots}
\pgfplotsset{compat=newest}
\usepackage{tikz}
\usepackage{tikzscale}
\usepackage{filecontents}
\usepackage[numbered]{bookmark}
\usepackage{cancel}
\usepackage{marvosym}
\usetikzlibrary{calc}
\newcommand\HRule{\rule{\textwidth}{1pt}}
\usepackage{multirow,array}
\usepackage{colortbl}

% This breaks long url's on several lines
\makeatletter
\g@addto@macro{\UrlBreaks}{\UrlOrds}
\makeatother

\DeclareRobustCommand\citepos
{\begingroup
	\let\NAT@nmfmt\NAT@posfmt% ...except with a different name format
	\NAT@swafalse\let\NAT@ctype\z@\NAT@partrue
	\@ifstar{\NAT@fulltrue\NAT@citetp}{\NAT@fullfalse\NAT@citetp}}

\pgfmathdeclarefunction{gauss}{2}{%
	\pgfmathparse{1/(#2*sqrt(2*pi))*exp(-((x-#1)^2)/(2*#2^2))}%
}

\usepackage{ragged2e}
\usepackage{etoolbox}
\usepackage{subfig}
\apptocmd{\frame}{}{\justifying}{} % Allow optional arguments after frame.
\setbeamertemplate{sidebar right}{}
\setbeamertemplate{footline}{\hfill\usebeamertemplate***{navigation symbols}}
\setbeamertemplate{frametitle}[default][center]
\usetheme{CambridgeUS} % Very good and clean


\begin{document}

	\begin{frame} %[allowframebreaks]
		\frametitle{Test}
		Test: \citet{smith1776wealth} \citet[p.~289]{smith1776wealth} and \citep{smith1776wealth} and \citep[p.~289]{smith1776wealth}.
	\end{frame}
		
	\begin{frame}[allowframebreaks]
		\frametitle{References}

		\bibliographystyle{apalike} 
		\bibliography{../JabRef2/RefLib
[Test_Cite_Beamer.pdf](https://github.com/josephwright/beamer/files/6171509/Test_Cite_Beamer.pdf)
}        
		
	\end{frame}
	
\end{document}

**Comment by simon-carlstedt** [comment]
```
%\documentclass{beamer}

\documentclass[
xcolor={svgnames},
hyperref={colorlinks=true,linkcolor=orange,citecolor=blue,urlcolor=blue,bookmarksopen=true}
]{beamer}
\mode<presentation>

\usepackage{amsmath}%
\usepackage{amsfonts}%
\usepackage{amssymb}%
\usepackage{graphicx}
\usepackage{amsthm}
\usepackage{mathrsfs}
\usepackage{array}
\usepackage{bm}
\usepackage{amsbsy}
\usepackage{tikz}
\usetikzlibrary{positioning,shapes,arrows}
\usepackage[square,comma,super,numbers,sort&compress]{natbib}
\usepackage{breakcites}
\usepackage{natbib}
\bibpunct{(}{)}{;}{a}{,}{,} % preamble for output, first two are bracket type, i.e. round
\setlength{\bibsep}{0pt} % Make reference list compact
\usepackage{tikz}
\usetikzlibrary{shapes,backgrounds,positioning,arrows}
\usetikzlibrary{chains,fit}
\usepackage{tikz}
\usepackage{verbatim}
\usepackage{rotating}
\usepackage{float}
\usepackage{pgfplots}
\pgfplotsset{compat=newest}
\usepackage{tikz}
\usepackage{tikzscale}
\usepackage{filecontents}
\usepackage[numbered]{bookmark}
\usepackage{cancel}
\usepackage{marvosym}
\usetikzlibrary{calc}
\newcommand\HRule{\rule{\textwidth}{1pt}}
\usepackage{multirow,array}
\usepackage{colortbl}

% This breaks long url's on several lines
\makeatletter
\g@addto@macro{\UrlBreaks}{\UrlOrds}
\makeatother

\DeclareRobustCommand\citepos
{\begingroup
	\let\NAT@nmfmt\NAT@posfmt% ...except with a different name format
	\NAT@swafalse\let\NAT@ctype\z@\NAT@partrue
	\@ifstar{\NAT@fulltrue\NAT@citetp}{\NAT@fullfalse\NAT@citetp}}

\pgfmathdeclarefunction{gauss}{2}{%
	\pgfmathparse{1/(#2*sqrt(2*pi))*exp(-((x-#1)^2)/(2*#2^2))}%
}

\usepackage{ragged2e}
\usepackage{etoolbox}
\usepackage{subfig}
\apptocmd{\frame}{}{\justifying}{} % Allow optional arguments after frame.
\setbeamertemplate{sidebar right}{}
\setbeamertemplate{footline}{\hfill\usebeamertemplate***{navigation symbols}}
\setbeamertemplate{frametitle}[default][center]
\usetheme{CambridgeUS} % Very good and clean


\begin{document}

	\begin{frame} %[allowframebreaks]
		\frametitle{Test}
		Test: \citet{smith1776wealth} \citet[p.~289]{smith1776wealth} and \citep{smith1776wealth} and \citep[p.~289]{smith1776wealth}.
	\end{frame}
		
	\begin{frame}[allowframebreaks]
		\frametitle{References}

		\bibliographystyle{apalike} 
		\bibliography{../JabRef2/ThesisIndMgt2}        
		
	\end{frame}
	

\end{document}
```

[Test_Cite_Beamer.pdf](https://github.com/josephwright/beamer/files/6171526/Test_Cite_Beamer.pdf)

**Comment by simon-carlstedt** [comment]
Thanks for all the great work on beamer btw :)

**Comment by samcarter** [comment]
A couple of comments about your code

- you don't need to bother with the various `\setbeamertemplate{...}{...}` if you overwrite them later by loading a theme
- you don't need `etoolbox`, `graphicx`, the `ams...` packages - beamer loads these automatically
- don't load the same package multiple times - you load some of them as often as four times
- instead of the `colortbl` package, use `\documentclass[xcolor=table]{beamer}`
- `filecontents` is deprecated 
- the `float` package does not make much sense with beamer

Now to your bib problem:

- I would not change the hyperref colours with beamer. Beamer sets them up very carefully and this will change colours at all kinds of weird places like the title in the footline
- `\bibpunct` is the old command for changing the natbib style and by using it you overwrite all the settings you made earlier during loading the package. Now I don't know if you actually intended the settings like square brackets, but from the pdf you show, I guess you could just use the `round` natbib option
- to add colour, you could use this quick hack (at least if you don't need the versions with two optional arguments of the cite commands):

```
\documentclass[
xcolor={svgnames,table},
]{beamer}

\usepackage[round]{natbib}

\let\oldcite=\cite
\let\oldcitet=\citet
\let\oldcitep=\citep 
\renewcommand{\citet}[2][]{\textcolor{blue}{\oldcitet[#1]{#2}}}
\renewcommand{\citep}[2][]{\textcolor{blue}{\oldcitep[#1]{#2}}}
\renewcommand{\cite}[2][]{\textcolor{blue}{\oldcite[#1]{#2}}}

\begin{document}

	\begin{frame} %[allowframebreaks]
		\frametitle{Test}
		Test: \citet{knuth} \citet[p.~289]{knuth} and \citep{knuth} and \citep[p.~289]{knuth} \cite{knuth}.
    
    test
	\end{frame}
		
	\begin{frame}[allowframebreaks]
		\frametitle{References}
		\bibliographystyle{apalike} 
		\bibliography{\jobname}        
	\end{frame}
	
\end{document}
```

![Screen Shot 2021-03-19 at 15 51 35](https://user-images.githubusercontent.com/43832342/111799153-0ba47d00-88cb-11eb-86cf-320636c4c440.png)

**Comment by simon-carlstedt** [comment]
Thanks for all the comments. And your hack works like a charm :) 

And sure, the preamble was just copy-pasted things when I added stuff to my presentations, so no thought behind it in terms of good coding; if it worked I left it, never bothered to clean it up (e.g., adding the same lib multiple times is obv. just idiotic). So thanks for comments on cleaning it up, deprecated libs etc.

For reference,  I attached the standard cite-look in article format. 

[Paper_single.pdf](https://github.com/josephwright/beamer/files/6172260/Paper_single.pdf)
